Hugging face trainer callback integration

parthraut opened the pull request:
Pull request for issue #24 : Draft for HuggingFace Global Power Limit Optimizer. Imports correctly locally, but needs to be further tested on machine with GPUs.

parthraut committed:
testing

parthraut committed:
draft of HuggingFace Compatible GPLO, added to power_limit.py

parthraut committed:
finished draft for HFGPLO, fixed import errors on usage

parthraut committed:
finished draft of HFGPLO, imports correctly but needs to be fully tested

jaywonchung commented on zeus/optimizer/__init__.py:
```diff
@@ -15,3 +15,4 @@
 """A collection of optimizers for various knobs."""
 
 from zeus.optimizer.power_limit import GlobalPowerLimitOptimizer
+from zeus.optimizer.power_limit import HFGPLO
```
Please run `bash scripts/lint.sh` to automatically format code and fix any lint errors. That's why CI is failing.

jaywonchung commented on zeus/optimizer/power_limit.py:
```diff
@@ -468,3 +470,150 @@ def _save_profile(self) -> None:
+
+    class Wrapper(TrainerCallback):
+        def __init__(self, *args, **kwargs) -> None:
+            self.plo = cls(*args, **kwargs)
```
Constructor signature binding missing

jaywonchung replied:
Also `self.plo` is not a good name, since we want the code here to be generic over optimizers. Maybe use `self.optimizer`.

jaywonchung commented on zeus/optimizer/power_limit.py:
```diff
@@ -468,3 +470,150 @@ def _save_profile(self) -> None:
+            model: PreTrainedModel,
+            **kwargs,
+        ) -> None:
+            self.plo.on_evaluate()  # what to set metric to?
```
Another TODO was to check the `TrainerState` of something else holds the current epoch's validation metric.

jaywonchung commented on zeus/optimizer/power_limit.py:
```diff
@@ -468,3 +470,150 @@ def _save_profile(self) -> None:
+    return Wrapper
+
+
+HFGPLO = make_hf(GlobalPowerLimitOptimizer)
```
Let's fix naming policies to `"HF" + cls.__name__`. The name you set in line 612 should be the same as the name you assign to in this line.

jaywonchung commented on zeus/run/dataloader.py:
```diff
@@ -1011,17 +1011,17 @@ def __next__(self):
-                        ] = energy_consumed[index]
+                        self.train_epoch_energy[index][self.epoch_num - 1] = (
+                            energy_consumed[index]
+                        )
```
Did `black` make this change? I'd rather not change unrelated stuff. Can you see if you can revert this change? Although if `black` insists then we'll have to comply and change.

parthraut replied:
yes, black made this change. I'll revert it and try to ignore it

parthraut replied:
reverted and passed lint

jaywonchung reviewed (changes requested):
Thanks for your work @parthraut! I did a quick review.

parthraut committed:
Rewrote HFGPLO to static class, wrote a simple test case (needs to be added to)

parthraut committed:
finished class, ready for PR

parthraut committed:
fixing formatting in dataloader

parthraut committed:
fixed accidental dependency on transformers

parthraut commented:
I simplified the class to only the methods necessary for HFGlobalPowerLimitOptimizer.

Testing of HFGlobalPowerLimitOptimizer is in test.py (I can remove it before merging). It tests 4 things:

1. The constructor signatures of GPLO and HFGPLO are exactly the same
2. HFGPLO inherits from TrainerCallback
3. HFGPLO can be used as a HuggingFace TrainerCallback. Single GPU Test
4. Ensure that HFGPLO can be used as a HuggingFace TrainerCallback. Multi GPU Test

The output of these tests seem to show that the GlobalPowerLimitOptimizer underneath is being called correctly. A snapshot is provided below:
<img width="873" alt="image" src="https://github.com/ml-energy/zeus/assets/68670266/84e819b5-edcd-42f9-9172-19935ae984fa">

Please let me know if there is anything else I should test or add to the code. Thanks!

jaywonchung commented on test.py:
You should not put random files in the root of the repository. We are using `pytest` for testing and those are kept under the `tests` directory. Also, tests should run without GPU to enable local machine development. I should have clarified these early on. I'll note these testing policies in our CONTRIBUTING.md.

Requested changes:
1. Separate out constructor signature tests and incorporate them into `tests/optimizer/test_power_limit_optimizer.py`.
2. Refactor the remaining part of `test.py` into `examples/huggingface`. One example training script that supports both single and multi GPU training, and a good README (probably copy and modify `examples/imagenet/README.md`).
3. Please order imports according to PEP8: https://peps.python.org/pep-0008/#imports

parthraut replied:
should there also be a test using HFGlobalPowerLimitOptimizer for single/multi GPU using HuggingFace transformers, or just the signature equivalence test? I'll add the example training script, but should there be a test along with it, testing the functionality of HFGlobalPowerLimitOptimizer?

jaywonchung replied:
If there's a nice way to test whether HF Trainer + global power limit optimizer actually calls all the required optimizer callbacks at the right moment it would be very nice. But if that's kinda tricky, I think we can just have the example training script and run it on our actual GPUs to check ourselves.

jaywonchung commented on test.py:
```diff
@@ -0,0 +1,89 @@
+            logging_steps=10,
+        )
+
+        trainer = Trainer(
```
How does Trainer determine the number of GPUs to use? I'm asking because this and line 79 are identical, whereas this is supposed to use one GPU whereas line 79 is supposed to use four. This should be taken care of in the example training script you will put under `examples/huggingface`.

jaywonchung commented on zeus/optimizer/__init__.py:
```diff
@@ -15,3 +15,4 @@
 """A collection of optimizers for various knobs."""
 
 from zeus.optimizer.power_limit import GlobalPowerLimitOptimizer
+from zeus.optimizer.power_limit import HFGlobalPowerLimitOptimizer
```
Merge import with above.

jaywonchung commented on zeus/optimizer/power_limit.py:
```diff
@@ -468,3 +472,93 @@ def _save_profile(self) -> None:
+        pl_step: int = 25,
+        profile_path: str | Path | None = None,
+    ) -> None:
+        r"""[Wrapped for Hugging Face Trainer Callback] Initialize the optimizer.
```
I think you can have the bracket just in the class docstring. Please remove from other methods.

jaywonchung commented on zeus/optimizer/power_limit.py:
```diff
@@ -468,3 +472,93 @@ def _save_profile(self) -> None:
+class HFGlobalPowerLimitOptimizer(TrainerCallback):
+    """[Wrapped for Hugging Face Trainer Callback] Optimizer for the power limit knob.
+
+    This optimizer uses the JIT profiling log to determine the optimal power limit.
```
```suggestion
    This optimizer uses the JIT profiling log to determine the optimal power limit.
    See [`GlobalPowerLimitOptimizer`][zeus.optimizer.power_limit.GlobalPowerLimitOptimizer]
    for the underlying optimizer implementation.
```

jaywonchung reviewed (changes requested):
Thanks for your work! Apart from specific comments, we need to document this change. Please take a look at the overall documentation at https://ml.energy and the doc source under `docs/`, and mention this in appropriate places. We want first time users who have no idea about how Zeus is organized to know that this exists.

parthraut committed:
Update zeus/optimizer/power_limit.py

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
finished huggingface example, fixed imports, added ctor test

parthraut committed:
fixed some styling issues

parthraut committed:
fixed minor formatting, ready for review

parthraut commented:
Made all the requested changes. I refactored the example, borrowed it from an example script from huggingface. Also updated the docs, but I was not sure whether to also add it to docs/index.md.

jaywonchung commented on examples/huggingface/README.md:
```diff
@@ -0,0 +1,61 @@
+# Integrating Zeus with HuggingFace 🤗
+
+This example will demonstrate how to integrate Zeus with `HuggingFace 🤗` using `HFGlobalPowerLimitOptimizer`.
```
```suggestion
This example will demonstrate how to integrate Zeus with `HuggingFace 🤗 Trainer` using `HFGlobalPowerLimitOptimizer`.
```
You can directly commit suggestion in the PR UI using the Commit suggestion button.

jaywonchung commented on examples/huggingface/README.md:
```diff
@@ -0,0 +1,61 @@
+
+This example will demonstrate how to integrate Zeus with `HuggingFace 🤗` using `HFGlobalPowerLimitOptimizer`.
+
+[`run_clm.py`](run_clm.py) was adapted from [HuggingFace 🤗's example training code for fine-tuning lanuguage models](https://github.com/huggingface/transformers/tree/main/examples/pytorch/language-modeling).
```
```suggestion
[`run_clm.py`](run_clm.py) was adapted from [HuggingFace 🤗's example training code for fine-tuning language models](https://github.com/huggingface/transformers/tree/f3aa7db439a2a3942f76c115197fe953984ac334/examples/pytorch/language-modeling).
```

jaywonchung commented on examples/huggingface/README.md:
```diff
@@ -0,0 +1,61 @@
+- [`HFGlobalPowerLimitOptimizer`](https://ml.energy/zeus/reference/optimizer/power_limit/#zeus.optimizer.power_limit.HFGlobalPowerLimitOptimizer): Online-profiles each power limit with `ZeusMonitor` and finds the cost-optimal power limit. Calls GlobalPowerLimitOptimizer under the hood.
+
+## Integration with HuggingFace 🤗 Trainer
+For easy use with [HuggingFace 🤗 Transformers](https://huggingface.co/docs/transformers/en/index), [`HFGlobalPowerLimitOptimizer`](zeus.optimizer.power_limit.HFGlobalPowerLimitOptimizer) is a drop-in compatible [HuggingFace 🤗 Trainer Callback](https://huggingface.co/docs/transformers/en/main_classes/callback). When initializing a [HuggingFace 🤗 Trainer](https://huggingface.co/docs/transformers/ma ...
```
I don't think these Zeus links will work

parthraut replied:
okay, should I remove them?

jaywonchung replied:
I meant this
```diff
-[`HFGlobalPowerLimitOptimizer`](zeus.optimizer.power_limit.HFGlobalPowerLimitOptimizer)
+[`HFGlobalPowerLimitOptimizer`](https://ml.energy/zeus/reference/optimizer/power_limit/#zeus.optimizer.power_limit.HFGlobalPowerLimitOptimizerzeus.optimizer.power_limit.HFGlobalPowerLimitOptimizer)
```

jaywonchung replied:
The `[zeus.optimizer.power_limit.HFGlobalPowerLimitOptimizer]` trick only works inside docstrings, which is enabled by `mkdocstrings` automatically parsing them and replacing them with links into the documentation site.

jaywonchung commented on examples/huggingface/run_clm.py:
```diff
@@ -0,0 +1,789 @@
+from zeus.optimizer import HFGlobalPowerLimitOptimizer
+
+# Will error if the minimal version of Transformers is not installed. Remove at your own risks.
+# check_min_version("4.38.0.dev0")
```
```suggestion
check_min_version("4.37.2")
```

I'm guessing you commented this out as a hack since 4.38.0 is not released yet and this script didn't run for you without installing transformers from source. This is bad practice, because someone who has an older version of transformers installed will try to run this and get an error, and think Zeus is bad (`pip install` will not automatically update an already-installed package).

I tested the script with transformers 4.37.2 and it works, so I suggested this change. However, in general, please don't implement hacks unless they have good justifications.

jaywonchung commented on examples/huggingface/run_clm.py:
```diff
@@ -0,0 +1,789 @@
+        metadata={
+            "help": 'The indices of the gpus to use for training (ex. "0,1,2"). If none, all gpus will be used.'
+        },
+    )
```
I tried out the script and giving `--gpu_indices "0"` still utilizes all four GPUs available for training but just limits the ZeusMonitor's monitoring scope to GPU 0. This is incorrect and confusing.

jaywonchung replied:
Okay, I looked up how to specify GPU indices for Trainer to use, and apparently it's just via the environment variable `CUDA_VISIBLE_DEVICES`. `ZeusMonitor` already supports that.

1. Remove `ZeusTrainingArguments`, both from code and README.
2. Simply instantiate as `ZeusMonitor()` without giving `gpu_indices`. This will automatically respect `CUDA_VISIBLE_DEVICES` and only monitor those devices.

jaywonchung commented on tests/optimizer/test_power_limit_optimizer.py:
```diff
@@ -38,6 +38,8 @@
     from pytest_mock import MockerFixture
 
+import inspect
+
```
Please order imports according to PEP8: https://peps.python.org/pep-0008/#imports
Ditto for all other files.

jaywonchung commented on tests/optimizer/test_power_limit_optimizer.py:
```diff
@@ -271,3 +273,10 @@ def test_power_limit_optimizer(
+    gplo_signature = inspect.signature(GlobalPowerLimitOptimizer.__init__)
+    hfgplo_signature = inspect.signature(HFGlobalPowerLimitOptimizer.__init__)
+
+    assert gplo_signature == hfgplo_signature, "GlobalPowerLimitOptimizer and HFGlobalPowerLimitOptimizer signatures do not match."
```
Files should end with a newline according to the UNIX specification, which is why GitHub is showing this red icon. There should be an End-Of-Line (EOL) setting in your editor.

jaywonchung commented on examples/huggingface/README.md:
````diff
@@ -0,0 +1,61 @@
+    --do_eval \
+    --output_dir /tmp/test-clm \
+    --gpu_indices="0,1,2,3" # Specify GPUs to ZeusMonitor. If left out, Zeus Monitor uses all available GPUs. 
+```
````
```suggestion
## Running the Example

By default, `Trainer` will make use of all available GPUs. If you would like to use only a subset of the GPUs, specify the `CUDA_VISIBLE_DEVICES` environment variable, which Zeus will also automatically respect.

```bash
python run_clm.py \
    --model_name_or_path gpt2 \
    --dataset_name wikitext \
    --dataset_config_name wikitext-2-raw-v1 \
    --per_device_train_batch_size 8 \
    --per_device_eval_batch_size 8 \
    --do_train \
    --do_eval \
    --output_dir /tmp/test-clm
```
```

jaywonchung commented on examples/huggingface/requirements.txt:
```diff
@@ -0,0 +1,8 @@
+protobuf
+evaluate
+scikit-learn
+transformers
```
```suggestion
transformers>=4.37.2
```

jaywonchung reviewed (changes requested):
Thanks for your work! I think the example, its README, and docs are nice and appropriate.

Most of the comments are nitpicks, which I am only doing because this is your first PR and I would like us to align coding standards. Please address them, and make those your standards for future PRs as well.

parthraut committed:
Update examples/huggingface/README.md

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
Update examples/huggingface/README.md

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
Update examples/huggingface/run_clm.py

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
Update examples/huggingface/README.md

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
Update examples/huggingface/requirements.txt

Co-authored-by: Jae-Won Chung <jwnchung@umich.edu>

parthraut committed:
fixed some style, removed broken links in example/huggingface/README.md

parthraut commented:
Updated the code with all suggestions. For the README.md in examples/huggingface, I was unsure what to do for the links to HFGlobalPowerLimitOptimizer, so I just removed them.

parthraut committed:
fixed links in readme

jaywonchung reviewed (approved):
LGTM! Thanks again for your great work!
